### PR TITLE
Cps/transonic assembly of velocity shape func with add nodes

### PR DESCRIPTION
**Pull request by eloisabaez**
Function to complete the assembly the highlighted vectors shown below to account for the additional nodes introduced by the upwind element.

Ultimately based on the definition of the residual in Section A.2.7 of Fully Simultaneous Coupling of the Full Potential Equation and the Integral Boundary Layer Equations in Three Dimensions by Brian Allen Nishida (MIT, 1996).

Screen Shot 2020-06-19 at 4 46 43 PM

### CODE REVIEW

**Comment by inigolcanalejo** [comment]
The implementation looks very good! I think it would be a good idea to make a small simple test with a current element and an upwind element for these two functions. Just to make sure we don't break anything in the future whenever we change something in the code. What do you think?

**Comment by eloisabaez** [comment]
Yes, I think it's a good idea. I just wasn't sure how to implement one as it's not possible to call the function from outside the element but maybe I can make application variables again. I made a test for myself (which just involved handing checking printing the variables) which can be seen in this branch: https://github.com/KratosMultiphysics/Kratos/tree/cps/transonic-assemble-velocity-shape-func-vects.

**Comment by inigolcanalejo** [comment]
True, I didnt realize they are private functuons. Actually as far as I see the functions do not access private members, so it might be better to have them in the utilities. In that way it is also easier to make a test.

**Comment by eloisabaez** [comment]
I think it's not possible to access the upwind element from the utilities though unless we put that into a application variable?

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]: I think it is better to use the equation Id rather than the node id. This is important for the nodes that have two degrees of freedom, like the trailing edge.

**Review by inigolcanalejo — DISMISSED** [review]
Hi Eloisa,

I see two things:

1. In order to access the upwind element one could pass it directly to the function. Not sure this is the way you would like to have it, I would suggest you merge this and once you finish the other functions and you see where you are calling this function we can think whether it can be moved or tested otherwise.

2. I realized that it would be better to use the equation id rather than the node id, because using the node id will fail for the nodes that have two dofs. This is not critical for the first cases you will run, cause the elements in touch with the trailing edge will be subsonic, but it will become relevant for the cases where the elements in touch with the trailing edge are also supersonic.

We have two options, we merge this and we test it in another pr or make the above changes in this pr. I'll let you decide.

**Comment by eloisabaez** [comment]
Hi Iñigo,

I switched the key to use the equation id instead of the geometry id. I will leave the upwind element questions for later though.

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]: I think we still have the same issue, I meant something like this:

```suggestion
    std::vector<size_t> upwind_element_ids;
    pGetUpwindElement()->EquationIdVector(upwind_element_ids, CurrentProcessInfo);
    
    for (int i = 0; i < TNumNodes; i++) { 
        auto current_id = std::find(current_element_ids.begin(), current_element_ids.end(),
                upwind_element_ids[i]);
```

Does it make sense?

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]: Here it is also better to use EquationIdVector, cause inside we already take care whether it is kutta or not

```suggestion
    this->EquationIdVector(current_element_ids, CurrentProcessInfo);
```

**Review by inigolcanalejo — CHANGES_REQUESTED** [review]
Thanks for making the changes, they are in the right direction. I have two more suggestions though.

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp
Comment (eloisabaez) [reply applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]: The issue I had with calling the EquationIdVector function was that it requires a non-const process info variable

Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]: I think it is still ok, I would pay de price of having a non const process info

**Review by inigolcanalejo — APPROVED** [review]

**Pull request merged**
